### DISCUSSION TITLE
How could I build a documentation generator using pyright?

### DISCUSSION THREAD

**Discussion by aspizu** [microsoft/pyright#4830] [Q&A]
I want to build a documentation generator that uses pyright's language server protocol. The documentation that we get when hovering on symbols in the code editor will be used.
So, it seems like I will require a LSP Client that will fetch the documentation. I haven't found such library for Python.

**Comment by erictraut** [comment]
The language server protocol is documented [here](https://microsoft.github.io/language-server-protocol/).

**Reply by aspizu** [reply]
is there a python library that is a LSP client?

**Reply by erictraut** [reply]
I'm not sure what you mean by that. The LSP protocol is, well, a protocol. You can write a client (or a server) in any language you want as long as it follows the protocol.

**Reply by aspizu** [reply]
yes but a easy to use client library would be better that wraps the api calls into functions.

**Reply by aspizu** [reply]
idk what im doing wrong here
```py
import json
from subprocess import PIPE, Popen
from typing import Any

process = Popen(["pyright-langserver", "--stdio"], stdin=PIPE, stdout=PIPE, stderr=PIPE)
payload = {}


def send(data: dict[str, Any]):
    content = json.dumps(
        {
            **data,
            "jsonrpc": "2.0",
            "id": 1,
            "method": "textDocument/completion",
            "params": {},
        }
    ).encode()
    payload = b"Content-Length: " + str(len(content)).encode() + b"\r\n\r\n" + content
    return process.communicate(input=payload)[0]


print(send({}))
```

**Reply by rchiodo** [reply]
This link might provide enough information to get you started:
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#lifeCycleMessages

It looks to me like you didn't send initialize first. The language server is waiting for that message before it does anything else.

**Reply by debonte** [reply]
If you're wanting to deliver this as a VS Code extension, you may be interested in the [Python Extension Template](https://code.visualstudio.com/api/advanced-topics/python-extension-template).

https://github.com/openlawlibrary/pygls may also be helpful.
